app/scripts/lib/rpc-method-middleware/handlers/send-metadata.ts

```diff
@@ -0,0 +1,84 @@
+import { ethErrors } from 'eth-rpc-errors';
+import type { SubjectType } from '@metamask/permission-controller';
+import type {
+  Json,
+  JsonRpcParams,
+  JsonRpcRequest,
+  JsonRpcResponse,
+} from '@metamask/utils';
+import type {
+  JsonRpcEngineEndCallback,
+  JsonRpcEngineNextCallback,
+} from '@metamask/json-rpc-engine';
+import { MESSAGE_TYPE } from '../../../../../shared/constants/app';
+import { AddSubjectMetadata, HandlerWrapperType } from './handlers-helper';
+
+type SendMetadataOptionsType = {
+  addSubjectMetadata: AddSubjectMetadata;
+  subjectType: SubjectType;
+};
+
+type SendMetadataType<
+  Params extends JsonRpcParams = JsonRpcParams,
+  Result extends Json = Json,
+> = {
+  implementation: (
+    _req: JsonRpcRequest<Params>,
+    res: JsonRpcResponse<Result>,
+    _next: JsonRpcEngineNextCallback,
+    end: JsonRpcEngineEndCallback,
+    { addSubjectMetadata, subjectType }: SendMetadataOptionsType,
+  ) => void;
+} & HandlerWrapperType;
+
+/**
+ * This internal method is used by our external provider to send metadata about
+ * permission subjects so that we can e.g. display a proper name and icon in
+ * our UI.
+ */
+
+const sendMetadata: SendMetadataType = {
+  methodNames: [MESSAGE_TYPE.SEND_METADATA],
+  implementation: sendMetadataHandler,
+  hookNames: {
+    addSubjectMetadata: true,
+    subjectType: true,
+  },
+};
```

Comment:
How about using the `satisfies` operator instead of a type annotation? `sendMetadata` is still validated against `SendMetadataType`, but the compiler is allowed to narrow `sendMetadata` so that e.g. the `true` literals here are preserved without being widened to `boolean`.

```suggestion
const sendMetadata = {
  methodNames: [MESSAGE_TYPE.SEND_METADATA],
  implementation: sendMetadataHandler,
  hookNames: {
    addSubjectMetadata: true,
    subjectType: true,
  },
} satisfies SendMetadataType;
```

Reply:
done